### DISCUSSION TITLE
AttributeError: 'str' object has no attribute 'branches'

### DISCUSSION THREAD

**Discussion by binbjz** [python-gitlab/python-gitlab#2004] [Q&A]
## Expected Behavior
project should be callable object

## Actual Behavior
AttributeError: 'str' object has no attribute 'branches'

## Specifications
  - python version: 3.10.4
  - python-gitlab version: 3.4.0
  - API version you are using (v3/v4): v4
  - Gitlab server version (or gitlab.com): GitLab Community Edition [13.2.4]

```py
def get_all_projects(gl: Gitlab, all_flag: bool = True) -> Generator:
    if not gl:
        raise Exception("Pls specify gl srv conn.")
    projects = gl.projects.list(all=all_flag, as_list=False, max_retries=12)
    return projects
```

```py
def cl_all_br_fmt_tmpl(prj: RESTObject, s_time: str, e_time: str, prj_lst: Optional[list] = None):
    if not prj_lst:
        prj_lst = []

    for branch in prj.branches.list(all=True):
        commits = prj.commits.list(all=True,
                                   query_parameters={"since": s_time,
                                                     "until": e_time,
                                                     "ref_name": branch.name})
        for commit in commits:
            if "Merge" not in commit.title:
                com = prj.commits.get(commit.id)
                _dict = {"committed_project_path": prj.path_with_namespace,
                         "committed_author": com.author_name,
                         "committed_branch": branch.name,
                         "committed_email": com.committer_email,
                         "committed_additions": com.stats["additions"],
                         "committed_deletions": com.stats["deletions"],
                         "committed_nums": com.stats["total"],
                         "committed_date": com.committed_date[:10]}
                prj_lst.append(_dict)
    return prj_lst
```

### First way - works fine (Multiple processes are not used)
```py
def cl_all_br_fmt(gl: Gitlab, s_time: str, e_time: str):
    if not gl:
        raise Exception("Pls specify gl srv conn.")

    projects = get_all_projects(gl)
        yield cl_all_br_fmt_tmpl(project, s_time, e_time)
```

### for branch in prj.branches.list(all=True): --> AttributeError: 'str' object has no attribute 'branches'
```py
def cl_all_br_fmt(gl: Gitlab, s_time: str, e_time: str):
    if not gl:
        raise Exception("Pls specify gl srv conn.")

    projects = get_all_projects(gl)
    pool_size = multiprocessing.cpu_count() * 3
    pool = multiprocessing.Pool(
        processes=pool_size,
    )
    p_func = partial(cl_all_br_fmt_tmpl, s_time, e_time)
    result_data = pool.map(p_func, projects)
    return result_data
```

### Execute the program to call cl_all_br_fmt
```py
gl = Gitlab(url=gl_url, private_token=gl_token, retry_transient_errors=True, timeout=30.0)
for itr in cl_all_br_fmt(gl, s_time, e_time):
    for i in itr:
        print(i)
```

### Why can the following method be called ？project.branches
```py
projects = gl.projects.list(all=True, as_list=False, max_retries=12)
for project in projects:
    for branch in project.branches.list(all=True):
```

### Please help me to fix this issue
```py
pool_size = multiprocessing.cpu_count() * 3
pool = multiprocessing.Pool(
    processes=pool_size,
)
p_func = partial(cl_all_br_fmt_tmpl, s_time, e_time)
projects = gl.projects.list(all=True, as_list=False, max_retries=12)
pool.map(p_func, projects)
```

**Comment by nejch** [comment]
Hi @binbjz, to me this is not really a python-gitlab issue, but how you're dealing with multiprocessing, so I'd try to find help elsewhere (or maybe in the community chat) - especially considering it works otherwise.

python-gitlab's requests to the server are I/O bound, so I don't think multiprocessing is the solution here, but asyncio:
https://stackoverflow.com/questions/27435284/multiprocessing-vs-multithreading-vs-asyncio-in-python-3

For an example of an async wrapper around python-gitlab, you can get some inspiration from https://github.com/pan-net-security/aio-gitlab.

So I think we can close this if that's ok but reopen if you disagree! :) Edit: or we can turn this into a discussion as well if the community wants to help.

**Comment by binbjz** [comment]
Hi @nejch, I really don't have any ideas on this part, but I was hoping to get some help. I don't understand why it's not callable. How do I use asynchrony for my program？

**Reply by nejch** [reply]
@binbjz I've converted this to a Q&A discussion for now. Please follow the links I posted above as a start which should have quite a few resources linked as well.
